platforms/core-configuration/kotlin-dsl-tooling-builders/src/main/kotlin/org/gradle/kotlin/dsl/tooling/builders/internal/IsolatedProjectsSafeKotlinDslScriptsModelBuilder.kt

```diff
@@ -0,0 +1,301 @@
+                additionalImports()
+            } ?: emptyList()
+
+        val classPath = (scriptClassPath + accessorsClassPath.bin)
+        val sourcePath = (gradleSource() + scriptSourcePath + accessorsClassPath.src)
+        val implicitImports = project.scriptImplicitImports + additionalImports
+        return IntermediateScriptModel(scriptFile, classPath, sourcePath, implicitImports)
+    }
+
+    private
+    fun gradleSource() =
+        SourcePathProvider.sourcePathFor(
+            scriptClassPath,
+            scriptFile,
+            project.rootDir,
+            project.gradle.gradleHomeDir,
+            SourceDistributionResolver(project)
+        )
+}
+
+
+internal
+interface IsolatedScriptsModelParameter {
+    val parentClassPath: ClassPath
+    val parentSourcePath: ClassPath
+}
+
+
+internal
+data class DefaultIsolatedScriptModelParameter(
+    override val parentClassPath: ClassPath,
+    override val parentSourcePath: ClassPath
+) : IsolatedScriptsModelParameter
+
+
+internal
+data class IsolatedScriptsModel(
+    val models: List<IntermediateScriptModel>
+)
+
+
+internal
+class IsolatedScriptsModelBuilder(
+    private val intermediateModelProvider: IntermediateToolingModelProvider
+) : ParameterizedToolingModelBuilder<IsolatedScriptsModelParameter> {
+
+    override fun getParameterType() = IsolatedScriptsModelParameter::class.java
+
+    override fun canBuild(modelName: String): Boolean =
+        modelName == "org.gradle.kotlin.dsl.tooling.builders.internal.IsolatedScriptsModel"
+
+    override fun buildAll(modelName: String, project: Project): IsolatedScriptsModel {
+        error("Calling $this without parameter is not supported")
+    }
+
+    override fun buildAll(modelName: String, parameter: IsolatedScriptsModelParameter, project: Project): IsolatedScriptsModel {
+        val models = projectHierarchyScriptModels(project as ProjectInternal, parameter, intermediateModelProvider)
+        return IsolatedScriptsModel(models)
+    }
+}
+
+
+private
+fun projectHierarchyScriptModels(
+    project: ProjectInternal,
+    parameter: IsolatedScriptsModelParameter,
+    intermediateModelProvider: IntermediateToolingModelProvider
+): List<IntermediateScriptModel> {
+
+    val models = mutableListOf<IntermediateScriptModel>()
+
+    // TODO: compute own classpaths
+//    scriptClassPath = project.scriptCompilationClassPath,
+//    scriptSourcePath = sourcePathFor(sourceLookupScriptHandlersFor(project))
+    val ownClassPath = parameter.parentClassPath
+    val ownSourcePath = parameter.parentSourcePath
+
+    models += listOfNotNull(project.discoverBuildScript()?.let {
+        buildScriptModelBuilder(it, project, ownClassPath, ownSourcePath).buildModel()
+    })
+
+    models += project.discoverPrecompiledScriptPluginScripts().map {
+        // TODO:isolated support fully
+        IntermediateScriptModel(it, parameter.parentClassPath, parameter.parentSourcePath, emptyList())
```

Comment:
🤔 `project.scriptImplicitImports` would already be a good approximation for precompiled script plugins

Reply:
By the way, `project.scriptImplicitImports` could be moved to the _common model_ at the very top since it would be a common prefix to all scripts. We don't need it here for isolated models, right?